src/main/java/pl/cleankod/exchange/provider/CurrencyConversionNbpService.java

```diff
@@ -17,10 +18,14 @@ public CurrencyConversionNbpService(ExchangeRatesNbpClient exchangeRatesNbpClien
     }
 
     @Override
+    @Cacheable(
+            cacheNames = "currency-cache",
+            key = "#money.amount().toPlainString().concat('-').concat(#money.currency()).concat('-').concat(#targetCurrency)"
+    )
     public Money convert(Money money, Currency targetCurrency) {
         RateWrapper rateWrapper = exchangeRatesNbpClient.fetch("A", targetCurrency.getCurrencyCode());
         BigDecimal midRate = rateWrapper.rates().get(0).mid();
-        BigDecimal calculatedRate = money.amount().divide(midRate, RoundingMode.HALF_UP);
+        BigDecimal calculatedRate = money.amount().divide(midRate, 2, RoundingMode.HALF_UP);
```

Comment:
Stable operations like the money calculations should be grouped together, thus it would be better to introduce another method like `exchange` which will know how to perform the calculation. Any other unstable logic should be pulled out from the low-level object to the upper layer e.g. to the domain services.